### ISSUE TITLE
ARM Support

### ISSUE THREAD

**Issue by tcirstea** [PyO3/maturin#237]
Any plans for supporting ARM? 

On a side note, building maturin works, but then fails to be used since this happens:
```
💥 maturin failed
Caused by: The architecture ARM is not supported
```

**Comment by konstin** [comment]
Do you have any documentation or reference wheel for native wheels on arm? From reading the distutils source code the architecture is `os.uname().machine`, but couldn't find a list of possible values or notes about python support of those architectures. You should also note that everything but x86 also isn't manylinux compatible, so it can't be published as wheel on pypi.

**Comment by programmerjake** [comment]
If you want, I can build some wheels on my armv7a computer.

**Comment by konstin** [comment]
A wheel as reference and a sysconfig (`python3 -m sysconfig`) would definitely be helpful!

**Comment by programmerjake** [comment]
see also: https://github.com/pypa/manylinux/issues/84

**Comment by programmerjake** [comment]
I created a wheel for gmpy2 since that doesn't take long.
[python3.5-wheel-ubuntu-16.04-on-odroid-xu4.tar.gz](https://github.com/PyO3/maturin/files/3837062/python3.5-wheel-ubuntu-16.04-on-odroid-xu4.tar.gz)

**Comment by konstin** [comment]
Thanks for those files! So the apparent differences are:
 * `linux_armv7l` instead of `linux_x86_64` for the wheel name
 * `arm-linux-gnueabihf` instead of `x86_64-linux-gnu` for the so name

Both travis ci and github actions support arm, so we could use either for continuous integration.

**Comment by programmerjake** [comment]
apparently cpython gets the platform name on non-android unix systems from https://github.com/python/cpython/blob/5c0c325453a175350e3c18ebb10cc10c37f9595c/configure.ac#L708-L849
pip gets wheel platform names from https://github.com/pypa/pip/blob/4447cb0fae4b38c12bb1cec2d59bb2a230638145/src/pip/_internal/pep425tags.py#L140-L164

see also [PEP 425](https://www.python.org/dev/peps/pep-0425/) and [PEP 427](https://www.python.org/dev/peps/pep-0427/)

**Comment by konstin** [comment]
I've tinkered a bit with this and while it's not that difficult to implement, the current `Target` struct and `OS` enum should be replaced by `Platform` enum to avoid this becoming to messy:

```rust
pub enum Target {
    Linux { manylinux: Manylinux, arch: Arch},
    Windows {is_64_bit: bool},
    Macos,
    FreeBSD,
}
```

`Target::current()` and `Target::from_target_triple()` should check for unsupported combinations and return understandable errors (and never panic as `from_target_triple` currently does).

**Comment by santteegt** [comment]
Hi,

Any updates on supporting ARM architectures? I'm getting `The architecture AARCH64 is not supported`

**Comment by pantunes** [comment]
Hi,

Any updates regarding this issue? I get the same error and I'm stuck.

A code fix or/and updates will be highly appreciated.

Thanks

**Comment by konstin** [comment]
It seems that only manylinux2014 added support for armv7l (https://www.python.org/dev/peps/pep-0599/#the-manylinux2014-policy), so it makes sense to require manylinux2014 here.

**Comment by ijl** [comment]
You can build #273 locally to try.

**Comment by konstin** [comment]
Fixed by #273 :tada:
